Get rid of seed concept in gossip

asias opened the pull request:
gossip: Get rid of seed concept
 
 The concept of seed and the different behaviour between seed nodes and
 non seed nodes generate a lot of confusion, complication and error for
 users. For example, how to add a seed node into into a cluster, how to
 promote a non seed node to a seed node, how to choose seeds node in
 multiple DC setup, edit config files for seeds, why seed node does not
 bootstrap.
 
 If we remove the concept of seed, it will get much easier for users.
 After this series, seed config option is only used once when a new node
 joins a cluster.
 
 Major changes:
 
 - Seed nodes are only used as the initial contact point nodes.
 
 - Seed nodes now perform bootstrap. The only exception is the first node
   in the cluster.
 
 - The unsafe auto_bootstrap option is now ignored.
 
 - Gossip shadow round now talks to all nodes instead of just seed nodes.
 
 Refs: #6845
 Tests: update_cluster_layout_tests.py + manual test

avikivity commented on gms/gossiper.cc:
```diff
@@ -783,31 +780,6 @@ void gossiper::run() {
-                /* Gossip to a seed if we did not do so above, or we have seen less nodes
-                   than there are seeds.  This prevents partitions where each group of nodes
-                   is only gossiping to a subset of the seeds.
-
```
How do we avoid this problem?

I don't think it's a problem that we really need to avoid, but let's make sure.

asias replied:
We avoid this problem by contacting all initial contact point nodes (seeds node currently).

avikivity replied:
Ah, this is just during startup. And we'll move to sending to all peers (if not bootstrapping), yes?

asias replied:
Yes.

avikivity commented:
Can you explain why you picked that algorithm? Why not pick N random nodes and talk to them?

In fact, why not pick one random node and talk to it, but do it at a rate N times faster than we did before?

asias commented:
> Can you explain why you picked that algorithm? 

The algorithm is inspired by the SWIM gossip implementation. It uses the shuffle and group mechanism.  I think it makes sense and it is better than what we have now. 

> Why not pick N random nodes and talk to them?

This is what we do right now. The downside is there is no guarantee all nodes are talked with in a bounded time.  In worst case, a small part of nodes are always chosen although this is very unlikely, but still. 
 
> In fact, why not pick one random node and talk to it, but do it at a rate N times faster than we did before?

Picking one random and increasing gossip round  still have the unbounded issue. In each gossip round, the heart beat of the local node is increased, it is not very useful to increase the hearbeat update frequency.

avikivity commented:
> > Can you explain why you picked that algorithm?
> 
> The algorithm is inspired by the SWIM gossip implementation. It uses the shuffle and group mechanism. I think it makes sense and it is better than what we have now.

Please add links in comments, both for future reference and for me to improve my understanding now.

> 
> > Why not pick N random nodes and talk to them?
> 
> This is what we do right now. The downside is there is no guarantee all nodes are talked with in a bounded time. In worst case, a small part of nodes are always chosen although this is very unlikely, but still.
> 
> > In fact, why not pick one random node and talk to it, but do it at a rate N times faster than we did before?
> 
> Picking one random and increasing gossip round still have the unbounded issue. In each gossip round, the heart beat of the local node is increased, it is not very useful to increase the hearbeat update frequency.

One thing we can do is decrease the probablity of choosing a node after we pick it. But I'm happy to pick an established algorithm and use it; just add the links please, and I'll review after studying it.

Can the other patch be applied directly? I think they're independent.

asias commented:
> > > Can you explain why you picked that algorithm?
> > 
> > 
> > The algorithm is inspired by the SWIM gossip implementation. It uses the shuffle and group mechanism. I think it makes sense and it is better than what we have now.
> 
> Please add links in comments, both for future reference and for me to improve my understanding now.

Here it is: https://www.cs.cornell.edu/projects/Quicksilver/public_pdfs/SWIM.pdf

> 
> > > Why not pick N random nodes and talk to them?
> > 
> > 
> > This is what we do right now. The downside is there is no guarantee all nodes are talked with in a bounded time. In worst case, a small part of nodes are always chosen although this is very unlikely, but still.
> > > In fact, why not pick one random node and talk to it, but do it at a rate N times faster than we did before?
> > 
> > 
> > Picking one random and increasing gossip round still have the unbounded issue. In each gossip round, the heart beat of the local node is increased, it is not very useful to increase the hearbeat update frequency.
> 
> One thing we can do is decrease the probablity of choosing a node after we pick it. But I'm happy to pick an established algorithm and use it; just add the links please, and I'll review after studying it.
> 
> Can the other patch be applied directly? I think they're independent.

asias commented:
> > > Can you explain why you picked that algorithm?
> > 
> > 
> > The algorithm is inspired by the SWIM gossip implementation. It uses the shuffle and group mechanism. I think it makes sense and it is better than what we have now.
> 
> Please add links in comments, both for future reference and for me to improve my understanding now.
> 
> > > Why not pick N random nodes and talk to them?
> > 
> > 
> > This is what we do right now. The downside is there is no guarantee all nodes are talked with in a bounded time. In worst case, a small part of nodes are always chosen although this is very unlikely, but still.
> > > In fact, why not pick one random node and talk to it, but do it at a rate N times faster than we did before?
> > 
> > 
> > Picking one random and increasing gossip round still have the unbounded issue. In each gossip round, the heart beat of the local node is increased, it is not very useful to increase the hearbeat update frequency.
> 
> One thing we can do is decrease the probablity of choosing a node after we pick it. But I'm happy to pick an established algorithm and use it; just add the links please, and I'll review after studying it.

OK.

> 
> Can the other patch be applied directly? I think they're independent.

Do you mean applying "gossip: Do not talk to seed node explicitly" without "gossip: Talk to live endpoints in a shuffled fashion"? The first one depends on the latter one.

avikivity commented:
> > Can the other patch be applied directly? I think they're independent.
> 
> Do you mean applying "gossip: Do not talk to seed node explicitly" without "gossip: Talk to live endpoints in a shuffled fashion"? The first one depends on the latter one.

Yes. Since they depend, I'll wait for an update.

asias commented:
> > > Can the other patch be applied directly? I think they're independent.
> > 
> > 
> > Do you mean applying "gossip: Do not talk to seed node explicitly" without "gossip: Talk to live endpoints in a shuffled fashion"? The first one depends on the latter one.
> 
> Yes. Since they depend, I'll wait for an update.

OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too. 

After that, we are almost done with the job of removing seed concept.

avikivity commented:
> OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> 
> After that, we are almost done with the job of removing seed concept.

Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).

avikivity commented:
> > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > After that, we are almost done with the job of removing seed concept.
> 
> Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).

Maybe not just for a test, but also for a short blog post explaining what the problems with seed nodes are, and this is a demonstration of how it works.

asias commented:
> > > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > > After that, we are almost done with the job of removing seed concept.
> > 
> > 
> > Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).
> 
> Maybe not just for a test, but also for a short blog post explaining what the problems with seed nodes are, and this is a demonstration of how it works.

OK.

asias commented:
> > > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > > After that, we are almost done with the job of removing seed concept.
> > 
> > 
> > Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).
> 
> Maybe not just for a test, but also for a short blog post explaining what the problems with seed nodes are, and this is a demonstration of how it works.



> > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > After that, we are almost done with the job of removing seed concept.
> 
> Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).

Sure. We are not there yet but we will.

asias commented:
Changes in a87fe0ca55c271fc99dd54430db27267b425d558:
- More tests was performed. 
- Added two more patches to get rid of the seed
  gossip: Introduce GOSSIP_GET_ENDPOINT_STATES verb
  gossip: Get rid of seed concept

avikivity commented:
@asias please add a description of the manual tests to the cover letter.

avikivity commented:
Also, you'll need to rebase since it looks like there are unrelated commits in there. Did you branch from next?

avikivity commented on gms/gossiper.cc:
```diff
@@ -737,36 +736,45 @@ void gossiper::run() {
+                if (_endpoints_to_talk_with.empty()) {
+                    std::shuffle(_live_endpoints.begin(), _live_endpoints.end(), _random_engine);
+                    // This guarantees the local node will talk with all nodes in live_endpoints once within nr_rounds gossip rounds.
+                    size_t nr_rounds = 10;
```
I asked for a link to material you based this on. Both for the review, and as in-code explanation of what is going on.

asias replied:
OK. I thought you just wanted the link in the PR comments.

avikivity commented on gms/gossiper.cc:
```diff
@@ -432,6 +432,24 @@ future<> gossiper::handle_shutdown_msg(inet_address from) {
+            } else {
+                it = apps.erase(it);
+            }
+        }
```
Why copy everything and then erase? Isn't it simpler to copy just what is wanted?

asias replied:
I wanted to avoid copying other members manually in endpoint_state.  But yes, we can copy just want we want. I forgot this one in the version I just sent. I will add it in the next round.

avikivity replied:
I see you are still copying.

asias replied:
Yes. I missed it in the version https://github.com/scylladb/scylla/pull/6848#issuecomment-662844040. I will fix it in a new version.

avikivity commented on gms/gossiper.cc:
```diff
@@ -557,6 +581,50 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
 }
 
+// Runs inside seastar::async context
+void gossiper::apply_state_locally_without_listener_notification(std::unordered_map<inet_address, endpoint_state> map) {
```
Does this not duplicate apply_state_locally()? Let's merge the implementations.

asias replied:
We can merge them. The downside is we will add a lot of if branches which makes apply_state_locally even harder to understand.

avikivity replied:
I'm sure it can be abstracted away. For example, a helper function `notify = [should_notify] (...) { if (should_notify) { ... } };` can make the branches invisible.

avikivity commented on message/messaging_service.hh:
```diff
@@ -409,6 +412,11 @@ public:
+    future<gms::gossip_get_endpoint_states_response> send_gossip_get_endpoint_states(msg_addr id, gms::gossip_get_endpoint_states_request request);
+
     // Wrapper for DEFINITIONS_UPDATE
     void register_definitions_update(std::function<rpc::no_wait_type (const rpc::client_info& cinfo, std::vector<frozen_mutation> fm,
```
Do we not need a new cluster feature to know we may send the new verb?

In general I prefer to avoid adding new verbs if not strictly needed, but I think this one is such a simplification it is worth it.

asias replied:
We can not use gossip features. This stuff is even before gossip starts

avikivity replied:
But, we can remember if we've seen the feature before or not.

OTOH, we'll only have the bad verb during upgrades to this version. Maybe we can tolerate it.

avikivity commented on message/messaging_service.cc:
```diff
@@ -1032,6 +1033,16 @@ future<> messaging_service::send_gossip_digest_ack2(msg_addr id, gossip_digest_a
+    return unregister_handler(messaging_verb::GOSSIP_GET_ENDPOINT_STATES);
+}
+future<gms::gossip_get_endpoint_states_response> messaging_service::send_gossip_get_endpoint_states(msg_addr id, gms::gossip_get_endpoint_states_request request) {
+    return send_message_timeout<future<gms::gossip_get_endpoint_states_response>>(this, messaging_verb::GOSSIP_GET_ENDPOINT_STATES, std::move(id), 5000ms, std::move(request));
```
It's better to expose the timeout. messaging_service shouldn't decide on what is the right timeout for gossip.

asias replied:
OK.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
     });
 }
```
changelog: s/talks to/attempts to talk to/

asias replied:
OK

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+                    logger.debug("Got get_endpoint_states response from {}", node);
+                    node_talked.insert(node);
+                    apply_state_locally_without_listener_notification(response.endpoint_state_map);
+                } catch (seastar::rpc::unknown_verb_error& e) {
```
It's better to avoid generating unknown rpc errors. They're frightening to users because they look like a malfunction.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+            for (const auto& node : nodes) {
+                logger.debug("Sent get_endpoint_states request to {}", node);
+                try {
+                    auto response = ms().send_gossip_get_endpoint_states(msg_addr(node), request).get0();
```
We should contact nodes in parallel. If we have 100 nodes in a remote data center with 100 ms latency, we get 10 seconds delay.

asias replied:
I wanted to get less contentions when applying the results. In most cases, the response from first node is the latest info, we will check and ignore the rest of 99 responses  by looking at the version. 

Since we can handle such parallel apply, talking to nodes in parallel has more advantages.  I will change it.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+            gms::application_state::HOST_ID,
+            gms::application_state::TOKENS,
+            gms::application_state::SUPPORTED_FEATURES}};
+        logger.info("The do_shadow_round started, nodes={}", nodes);
```
Q: how do we learn about supported features of nodes that are down? without that, we cannot join a cluster that has a node down.

asias replied:
We learn them from the system  peer  table. So the features check is a combination of gossip live data + data from system tables.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+            }
+            if (!node_talked.empty()) {
+                break;
+            }
```
It's nicer to put this in the top-level loop condition, so readers see what the loop is trying to achieve.

asias replied:
OK. We still have other break condition in the loop though.

avikivity replied:
Didn't see this move. The condition below can also move.

avikivity commented on service/storage_service.cc:
```diff
@@ -205,12 +205,36 @@ bool get_property_load_ring_state() {
+    }
+    // Node with the smallest IP address is chosen as the very first node
+    // in the cluster. The first node is the only node that does not
+    // bootstrap in the cluser. All other nodes will bootstrap.
```
Good. Since it's illegal to bootstrap a cluster with different versions we won't run into different decisions by different nodes.

The only issue is if different nodes have different seed lists. We must document against it.

asias replied:
Yes. We need to document and enforce this.

asias commented:
> > > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > > After that, we are almost done with the job of removing seed concept.
> > 
> > 
> > Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).
> 
> Maybe not just for a test, but also for a short blog post explaining what the problems with seed nodes are, and this is a demonstration of how it works.



> > OK. In the mean while, I am looking at gossip shadow round stuff which depends on the seed concept. My idea is to replace "talk to seeds and wait for one to respond and proceed" with "talk to all nodes (all existing node for existing nodes, initial contact nodes for new nodes) and wait for them to respond". It is more robust and also removes the dependence to use seeds too.
> > After that, we are almost done with the job of removing seed concept.
> 
> Good. I suggest as a test, start a cluster with two seed nodes, and bootstrap and decommission until all seed nodes are gone, then bootstrap a node (we can still use the seed specification as the initial contact point).

Sure. We are not there yet but we will. 

> @asias please add a description of the manual tests to the cover letter.

OK.

avikivity commented on service/storage_service.cc:
```diff
@@ -232,62 +256,20 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
-                        _gossiper.add_saved_endpoint(ep);
-                    }
-                }
-            }
```
How does this removal work in a mixed environment before all nodes are upgraded?

asias replied:
For a mixed cluster, the shadow round will fallback to the old method.  It is a upgrade operation, so this is going to be an exiting cluster. The check should just work.

asias commented:
Changes in 3c11d38de48206b43380dd8bc7f48da9421e738c:
-  Change "talks to" to  "attempts to talk" in commit log. 
-  Expose the timeout for send_gossip_get_endpoint_states.
-  Send send_gossip_get_endpoint_states in parallel. 
-  Mention SWIM in code comments. 
-  Add manual test description in commit log.
-  Rebase to master.

asias commented:
Changes in 87d24a196fe271eb18912c0de587a992462419e9:
-  Change "talks to" to  "attempts to talk" in commit log. 
-  Expose the timeout for send_gossip_get_endpoint_states.
-  Send send_gossip_get_endpoint_states in parallel. 
-  Mention SWIM in code comments. 
-  Add manual test description in commit log.
-  Rebase to master.

asias commented:
> Also, you'll need to rebase since it looks like there are unrelated commits in there. Did you branch from next?

Yes, I rebased to latest master now.

avikivity commented on gms/gossiper.cc:
```diff
@@ -582,49 +582,84 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
-                logger.info("Applying remote_state for node {} (remote generation > local generation)", node);
+void gossiper::do_apply_state_locally(gms::inet_address node, const endpoint_state& remote_state, bool listener_notification) {
+    // If state does not exist just add it. If it does then add it if the remote generation is greater.
+    // If there is a generation tie, attempt to break it by heartbeat version.
```
It's impossible to review this large code chunks. Can you just change apply_state_locally() directly (perhaps renaming it to do_apply_state_locally with an extra parameter).

asias replied:
do_apply_state_locally extracts the common code. It is much easier to review without touching the existing apply_state_locally. But you suggested to merge them which unavoidably makes the patch set larger to review. 

The code t is not that complicated: it is like compare generation hen compare version.

avikivity replied:
For me it's harder. There are many lines moved, and I can't detect what the incremental changes are.

asias replied:
The code inside listener_notification branch is existing code and the one inside !listener_notification is new code.

avikivity replied:
Maybe I'm missing something. I saw big code movements. Maybe you forgot to rebase or push?

asias replied:
The move was splitting common code from apply_state_locally to do_apply_state_locally. I squashed it in new version: https://github.com/scylladb/scylla/pull/6848#issuecomment-663314943. Hope it is easier to review now.

avikivity replied:
It would be easier if you first create do_apply_state_locally() without the listener_notification flag, which is a pure code movement. Then the this patch adds listener_notification and it stands out from the code movement in the previous patch.

asias replied:
OK. I added a version do_apply_state_locally without listener_notification.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1761,38 +1761,83 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
     });
 }
```
Please also test upgrades and downgrades from pre-seedless.

avikivity replied:
@roydahan QA should prepare for this.

roydahan replied:
I agree, let's run dtest BYO even before merging this because it can break even next-gating tests.
We will start making sure that SCT will be ready for it.

asias replied:
I started a byo test but the test failed:

```
Error Message

unsupported url or file doesn't exist
	url=/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla-tools-java/build/scylla-tools-package.tar.gz
```
E.g., 

https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/986/testReport/junit/rebuild_test/TestRebuild/dtest_not_Parallel___simple_rebuild_test/

I changed the scylla git url and branch to test. All others are default. What else should I configure to make the byo test run?

asias commented:
Changes in 805ca394bba310325d477f16a2da33fe0484b6ca:
- Squash "gossip: Add do_apply_state_locally helper" into "gossip: Introduce GOSSIP_GET_ENDPOINT_STATES verb" to make the change easier to review
- Avoid coping in gossiper::handle_get_endpoint_states_msg

avikivity commented on gms/gossiper.cc:
```diff
@@ -557,6 +583,85 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
+        if (node == this->get_broadcast_address()) {
+            continue;
+        }
+        do_apply_state_locally(node, remote_state, false);
```
When do we notify the listeners? Can't we create a situation where we learn something from the cluster, but lost the infomation because we don't notify?

asias replied:
This is for shadow round only. After gossip service is started normally, we will learn everything and do the notification properly.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1763,38 +1763,83 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+        bool fall_back_to_syn_msg = false;
+        std::list<gms::gossip_get_endpoint_states_response> responses;
+        for (;;) {
+            parallel_for_each(nodes.begin(), nodes.end(), [this, &request, &responses, &nodes_talked, &nodes_down, &fall_back_to_syn_msg] (gms::inet_address node) {
```
Should we talk to a node more than once?

asias replied:
It makes sense to talk to a node again after sleep and retry, because it might have newer information since the last try. The last try did not succeed is probably due to the cluster is not stable yet in the first place.

avikivity commented:
Looks good. Please manually test and document cluster upgrade (do we need to test seed node upgraded first, and seed node upgraded last separately?)

What are the results of dtest?

asias commented:
> Looks good. Please manually test and document cluster upgrade (do we need to test seed node upgraded first, and seed node upgraded last separately?)

I think it is unnecessary to test upgrade staring from seed vs starting from non seed. Existing nodes will attempt to talk to all peer nodes anyway. 

I still tried testing both just to be sure. I added the upgrade doc and tests in the commit log.  


> What are the results of dtest?

I tried BYO but the test did not really run due to config errors. 

The dtest of update_cluster_layout_tests.py which has tons of start and stop procedures  passed locally.

asias commented:
Changes in 19cd605f1b259562d1dd4f911e4be05e383f26ee: 
- Split do_apply_state_locally with a version that does not have notification flag.
- Added upgrade and downgrade tests
- Added upgrade procedure 
- Added the missing fall_back_to_syn_msg break in the loop

avikivity commented:
Thanks. There is now blockage in next promotion, so I'm not queuing any risky stuff until it unblocks. In the meanwhile, please try to get byo working with @hagitsegev and @roydahan.

asias commented:
> Thanks. There is now blockage in next promotion, so I'm not queuing any risky stuff until it unblocks. In the meanwhile, please try to get byo working with @hagitsegev and @roydahan.

Could you start a BYO job for me, @hagitsegev / @roydahan ? My branch is at github.com:scylladb/scylla-dev.git gossip_no_seed. Other than the branch, anything else I should provide?

amoskong commented:
> > Thanks. There is now blockage in next promotion, so I'm not queuing any risky stuff until it unblocks. In the meanwhile, please try to get byo working with @hagitsegev and @roydahan.
> 
> Could you start a BYO job for me, @hagitsegev / @roydahan ? My branch is at github.com:scylladb/scylla-dev.git gossip_no_seed. Other than the branch, anything else I should provide?

Roy asked me to work it, I will try to build the packages by myself.

asias commented:
> > > Thanks. There is now blockage in next promotion, so I'm not queuing any risky stuff until it unblocks. In the meanwhile, please try to get byo working with @hagitsegev and @roydahan.
> > 
> > 
> > Could you start a BYO job for me, @hagitsegev / @roydahan ? My branch is at github.com:scylladb/scylla-dev.git gossip_no_seed. Other than the branch, anything else I should provide?
> 
> Roy asked me to work it, I will try to build the packages by myself.

Hello Amos, any luck with the byo test?

amoskong commented:
I still didn't make the job run successfully

On Fri, Jul 31, 2020 at 10:28 AM Asias He <notifications@github.com> wrote:

> Thanks. There is now blockage in next promotion, so I'm not queuing any
> risky stuff until it unblocks. In the meanwhile, please try to get byo
> working with @hagitsegev <https://github.com/hagitsegev> and @roydahan
> <https://github.com/roydahan>.
>
> Could you start a BYO job for me, @hagitsegev
> <https://github.com/hagitsegev> / @roydahan <https://github.com/roydahan>
> ? My branch is at github.com:scylladb/scylla-dev.git gossip_no_seed.
> Other than the branch, anything else I should provide?
>
> Roy asked me to work it, I will try to build the packages by myself.
>
> Hello Amos, any luck with the byo test?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/6848#issuecomment-666877549>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACLTTH55K64VWJT5YD74RLR6IT4BANCNFSM4O2M7YNQ>
> .
>

bhalevy commented:
> I still didn't make the job run successfully

Let's open an issue in github about it, so we can track the fix

avikivity commented:
@amoskong @penberg @roydahan is there any way to run the test? this series really wants to be committed.

asias commented:
> @amoskong @penberg @roydahan is there any way to run the test? this series really wants to be committed.

Ping. Hello folks, any progress? What should we proceed?

bhalevy commented:
@avikivity would a byo run of dtest-release be enough?

avikivity commented:
@bhalevy yes

bhalevy commented:
@asias I'm trying the gating tests over your branch (rebased onto master 74db08165ddcc2db061b88b4a28e0fa682305655) here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1027

asias commented:
> @asias I'm trying the gating tests over your branch (rebased onto master [74db081](https://github.com/scylladb/scylla/commit/74db08165ddcc2db061b88b4a28e0fa682305655)) here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/
> (Edit, run 1027 failed due to power outage)

Thanks, benny. It looks like we still have issue with the test itself. A lot of tests (532) failed.

https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/#showFailuresLink

bhalevy commented:
> > @asias I'm trying the gating tests over your branch (rebased onto master [74db081](https://github.com/scylladb/scylla/commit/74db08165ddcc2db061b88b4a28e0fa682305655)) here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/
> > (Edit, run 1027 failed due to power outage)
> 
> Thanks, benny. It looks like we still have issue with the test itself. A lot of tests (532) failed.
> 
> https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/#showFailuresLink

@amoskong did you this kind of failures in unit tests in byo?
I think it's unlikely that the address in use issue is related to @asias's changes.

In any case the gating dtest passed:
```
21:49:43 Ran 183 tests in 10343.758s
21:49:43 OK (SKIP=8)
```

asias commented:
> > > @asias I'm trying the gating tests over your branch (rebased onto master [74db081](https://github.com/scylladb/scylla/commit/74db08165ddcc2db061b88b4a28e0fa682305655)) here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/
> > > (Edit, run 1027 failed due to power outage)
> > 
> > 
> > Thanks, benny. It looks like we still have issue with the test itself. A lot of tests (532) failed.
> > https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/#showFailuresLink
> 
> @amoskong did you this kind of failures in unit tests in byo?
> I think it's unlikely that the address in use issue is related to @asias's changes.
> 
> In any case the gating dtest passed:
> 
> ```
> 21:49:43 Ran 183 tests in 10343.758s
> 21:49:43 OK (SKIP=8)
> ```

Thanks Benny. 

@avikivity Do you need anything else to be tested before we merge it?

amoskong commented:
> > > @asias I'm trying the gating tests over your branch (rebased onto master [74db081](https://github.com/scylladb/scylla/commit/74db08165ddcc2db061b88b4a28e0fa682305655)) here: https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/
> > > (Edit, run 1027 failed due to power outage)
> > 
> > 
> > Thanks, benny. It looks like we still have issue with the test itself. A lot of tests (532) failed.
> > https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/byo_build_tests_dtest/1028/#showFailuresLink
> 
> @amoskong did you this kind of failures in unit tests in byo?
> I think it's unlikely that the address in use issue is related to @asias's changes.

I saw more failure tests in another byo:
https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1029/

> In any case the gating dtest passed:
> ```
> 21:49:43 Ran 183 tests in 10343.758s
> 21:49:43 OK (SKIP=8)
> ```

avikivity commented:
@asias what about the failures @amoskong reports?

asias commented:
> @asias what about the failures @amoskong reports?

I do not think they are relevant.

```
  | Test Result (983 failures  / +451)test/boost/filtering_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_allow_filtering_checktest/boost/filtering_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_allow_filtering_pk_cktest/boost/filtering_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_allow_filtering_multi_columntest/boost/filtering_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_filtering_on_empty_partition_with_a_static_rowtest/boost/filtering_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_filteringtest/boost/json_cql_query_test.cc./data/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/src/testing/seastar_test.test_select_json_types
-- | --


```

bhalevy commented:
> > @asias what about the failures @amoskong reports?
> 
> I do not think they are relevant.
> 

I agree.  To be absolutely sure, and as a basis for opening a github issue about it,
I launched https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1034 that will run the unit tests on a pristine `next` build.

avikivity commented:
:crossed_fingers:

avikivity commented:
Dequeueing, almost all unit tests fail.

asias commented:
On Sun, Aug 16, 2020 at 1:27 AM Avi Kivity <notifications@github.com> wrote:

> Dequeueing, almost all unit tests fail.
>

It turned out to

unknown location(0): fatal error: in
"test_partition_key_compound_restrictions": std::system_error: posix_listen
failed for address 127.0.0.1:7000: Address already in use

I fixed it by respecting the do_bind option in do_shadow_round too.

bash-5.0# ./test.py
================================================================================
[N/TOTAL] TEST                                                 MODE
RESULT
------------------------------------------------------------------------------

[149/149] boost/cql_query_test                            dev    [ PASS ]

------------------------------------------------------------------------------

Here is the new branch:

git@github.com:asias/scylla.git gossip_no_seed_v2


-- 
Asias

avikivity commented:
Are you sending a new version?

asias commented:
> Are you sending a new version?

I already did. (It was hidden in the bottom of the last comments.)

Here is the new branch:
git@github.com:asias/scylla.git gossip_no_seed_v2

avikivity commented:
Don't forget the blog post. We'll need to educate users about it.

amoskong commented:
> > Are you sending a new version?
> 
> I already did. (It was hidden in the bottom of the last comments.)
> 
> Here is the new branch:
> [git@github.com](mailto:git@github.com):asias/scylla.git gossip_no_seed_v2

I had copied this PR to git@github.com:scylladb/scylla-dev.git (amos/gossip_no_seed_v2 branch).
And executed a byo job: 
- https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1035/testReport/
- version: 666.development-0.20200817.d0b3f3dfe80  (matches with Asias' branch)

```
Test Result
2 failures (-517) , 55 skipped (-12)
2,418 tests (-29)
Took 3 hr 6 min.
```

Only two subtests of migration failed, but the error doesn't relate with Asias' change.

```
bin/nodetool -h 127.0.38.1 -p 7138 refresh -- ks cf' failed; exit status: 1; stdout: nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/ks' failed: Failed to load new sstables: sstables::malformed_sstable_exception (invalid version for file md-2-big-Index.db. Name doesn't match any known version.)
```

asias commented:
> > > Are you sending a new version?
> > 
> > 
> > I already did. (It was hidden in the bottom of the last comments.)
> > Here is the new branch:
> > [git@github.com](mailto:git@github.com):asias/scylla.git gossip_no_seed_v2
> 
> I had copied this PR to [git@github.com](mailto:git@github.com):scylladb/scylla-dev.git (amos/gossip_no_seed_v2 branch).
> And executed a byo job:
> 
>     * https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1035/testReport/
> 
>     * version: 666.development-0.20200817.d0b3f3dfe80  (matches with Asias' branch)
> 
> 
> ```
> Test Result
> 2 failures (-517) , 55 skipped (-12)
> 2,418 tests (-29)
> Took 3 hr 6 min.
> ```
> 
> Only two subtests of migration failed, but the error doesn't relate with Asias' change.
> 
> ```
> bin/nodetool -h 127.0.38.1 -p 7138 refresh -- ks cf' failed; exit status: 1; stdout: nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/ks' failed: Failed to load new sstables: sstables::malformed_sstable_exception (invalid version for file md-2-big-Index.db. Name doesn't match any known version.)
> ```

Thanks, Amos!

asias commented:
> Don't forget the blog post. We'll need to educate users about it.

Yes, in my queue, after the release issues.

avikivity commented:
> Only two subtests of migration failed, but the error doesn't relate with Asias' change.
> 
> ```
> bin/nodetool -h 127.0.38.1 -p 7138 refresh -- ks cf' failed; exit status: 1; stdout: nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/ks' failed: Failed to load new sstables: sstables::malformed_sstable_exception (invalid version for file md-2-big-Index.db. Name doesn't match any known version.)
> ```


Looks like the test needs to be adjusted for md format.

bhalevy commented:
> > Only two subtests of migration failed, but the error doesn't relate with Asias' change.
> > ```
> > bin/nodetool -h 127.0.38.1 -p 7138 refresh -- ks cf' failed; exit status: 1; stdout: nodetool: Scylla API server HTTP POST to URL '/storage_service/sstables/ks' failed: Failed to load new sstables: sstables::malformed_sstable_exception (invalid version for file md-2-big-Index.db. Name doesn't match any known version.)
> > ```
> 
> Looks like the test needs to be adjusted for md format.

FWIW they pass in master ee7291aa882d925d3de3dd3db14df443db020034
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/579/testReport/migration_test/TestMigrationUpgradeSSTables_with_3_0_md/migrate_sstable_with_counter_test/
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/579/testReport/migration_test/TestMigration_with_3_0_md/migrate_sstable_with_counter_test/
